File: android/src/main/java/expo/modules/staticserver/ExpoStaticServerModule.kt
package expo.modules.staticserver

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import fi.iki.elonen.NanoHTTPD
import java.io.IOException

class ExpoStaticServerModule : Module() {
  private lateinit var httpServer: HttpServer

  override fun definition() = ModuleDefinition {
    Name("ExpoStaticServer")

    AsyncFunction("startServer") { host: String, port: Int, root: String, promise: Promise ->

      try{
        if(::httpServer.isInitialized) {
          httpServer.stop()
        }
        httpServer = HttpServer(host, port, root)
        httpServer.start(NanoHTTPD.SOCKET_READ_TIMEOUT, true)
        promise.resolve("ok")
      } catch(e: IOException) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("stopServer") { promise: Promise ->
      try {
        if(::httpServer.isInitialized) {
          httpServer.stop()
        }
        promise.resolve("ok")
      } catch(e: IOException) {
        promise.reject(CodedException(e))
      }
    }
  }
}

File: ios/ExpoStaticServerModule.swift
import ExpoModulesCore
import GCDWebServer

public class ExpoStaticServerModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  private var httpServer: HttpServer?
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoStaticServer')` in JavaScript.
    Name("ExpoStaticServer")

    AsyncFunction("startServer") { (host: String, port: UInt, root: String, promise: Promise) in
        DispatchQueue.main.async {
            do {
                if let server = self.httpServer {
                  server.stop()
                  self.httpServer = nil
                }
                self.httpServer = try HttpServer(host: host, port: port, root: root)
                promise.resolve("ok")
            } catch let error {
                promise.reject("START_SERVER_ERROR", error.localizedDescription)
            }
        }
    }

    AsyncFunction("stopServer") { (promise: Promise) in
        if let server = self.httpServer {
          server.stop()
          self.httpServer = nil
        }
        promise.resolve("ok")
    }
  }
}
